### PR TITLE
READY: Implemented endpoint to start a download from infohash

### PR DESCRIPTION
**Pull request by devos50**

### CODE REVIEW

**Comment by devos50** [comment]
(test seems to hang?)

**Comment by devos50** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by devos50** [comment]
@lfdversluis ready for your review

**Comment by lfdversluis** [comment]
Looks good to me.

**Comment by devos50** [comment]
@lfdversluis I'm thinking about adding some parameters for anonymous downloading/seeding + destination on disk so I will work on that first.

**Comment by brussee** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
See #2363

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
Tests on mac timed out

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
The error encountered here is probably fixed by #2392

**Comment by devos50** [comment]
retest this please

File: Tribler/Core/Modules/restapi/downloads_endpoint.py [hunk 1]
```diff
@@ -134,6 +137,35 @@ def __init__(self, session, infohash):
...
+        """
+        Create a download configuration based on some given parameters. Possible parameters are:
+        - anon_download: whether the download is anonymous or not
+        - safe_seeding: whether seeding of the torrent is anonymous
+        - destination: the destination path of the torrent (where it is saved on disk)
+        """
+        download_config = DownloadStartupConfig()
+
+        has_anon_download = 'anon_download' in parameters and len(parameters['anon_download']) > 0
+        has_safe_seeding = 'safe_seeding' in parameters and len(parameters['safe_seeding']) > 0
```
Comment (devos50) [inline Tribler/Core/Modules/restapi/downloads_endpoint.py]: (Laurens): If now someone specifies (anon_download = false), then your code will think it's true. This is more a flag based approach than an argument based approach.

File: Tribler/Core/Modules/restapi/downloads_endpoint.py [hunk 2]
```diff
@@ -134,6 +137,35 @@ def __init__(self, session, infohash):
...
+
+        if not has_anon_download and has_safe_seeding:
+            return None, "Cannot set safe_seeding without anonymous download enabled"
+
+        if has_anon_download:
+            download_config.set_hops(1)
```
Comment (devos50) [inline Tribler/Core/Modules/restapi/downloads_endpoint.py]: (Laurens): Can't the number of hops be an argument too? Default 1?

File: Tribler/Core/Modules/restapi/downloads_endpoint.py (hunk 1 above)
Comment (devos50) [inline Tribler/Core/Modules/restapi/downloads_endpoint.py]: @whirm do you know the best way to make it a flag?

Comment (whirm) [inline Tribler/Core/Modules/restapi/downloads_endpoint.py]: @devos50 I'll come to your office later to talk about this

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
Job got aborted halfway?

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
```
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/tribler/Documents/workspace/workspace/GH_Tribler_PR_tests_mac/tribler/Tribler/Core/Utilities/twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/local/lib/python2.7/site-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/local/lib/python2.7/site-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/local/lib/python2.7/site-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/Users/tribler/Documents/workspace/workspace/GH_Tribler_PR_tests_mac/tribler/Tribler/Core/Libtorrent/__init__.py", line 27, in <lambda>
    lambda_f = lambda a = args, kwa = kwargs: invoke_func(*a, **kwa)
  File "/Users/tribler/Documents/workspace/workspace/GH_Tribler_PR_tests_mac/tribler/Tribler/Core/Libtorrent/__init__.py", line 28, in invoke_func
    download.session.lm.threadpool.add_task(lambda_f, 1)
exceptions.AttributeError: 'NoneType' object has no attribute 'add_task'
```

**Comment by devos50** [comment]
retest this please

**Comment by brussee** [comment]
retest this please

**Comment by devos50** [comment]
```
test_ffsearch (Tribler.Test.test_remote_search.TestMixedRemoteSearch) ... Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/Utilities/twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "<string>", line 2, in _do_request
    
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/RemoteTorrentHandler.py", line 32, in pass_when_stopped
    return f(self, *argv, **kwargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/RemoteTorrentHandler.py", line 455, in _do_request
    trackers = self._torrent_db_handler.getTrackerListByInfohash(infohash)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/SqliteCacheDBHandler.py", line 696, in getTrackerListByInfohash
    torrent_id = self.getTorrentID(infohash)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/SqliteCacheDBHandler.py", line 242, in getTorrentID
    return self.getTorrentIDS([infohash, ]).get(infohash)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/SqliteCacheDBHandler.py", line 262, in getTorrentIDS
    torrents = self._db.fetchall(sql_stmt, to_select)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/dispersy/util.py", line 46, in helper
    return blockingCallFromThread(reactor, func, *args, **kargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/dispersy/util.py", line 225, in blockingCallFromThread
    return f(*args, **kwargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/sqlitecachedb.py", line 401, in fetchall
    res = self.execute_read(sql, args)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/sqlitecachedb.py", line 310, in execute_read
    return self.execute(sql, args)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/dispersy/util.py", line 46, in helper
    return blockingCallFromThread(reactor, func, *args, **kargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/dispersy/util.py", line 225, in blockingCallFromThread
    return f(*args, **kwargs)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/sqlitecachedb.py", line 263, in execute
    cur = self.get_cursor()
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux@2/tribler/Tribler/Core/CacheDB/sqlitecachedb.py", line 196, in get_cursor
    self._cursor_table[thread_name] = self._connection.cursor()
exceptions.AttributeError: 'NoneType' object has no attribute 'cursor'
```

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
This one is too unstable too be merged at the moment due to the thread pool stuff when checkpointing (this is probably since I'm starting downloads in the tests I've added).

**Comment by whirm** [comment]
I'm working on removing the threadpool altogether, but I'm finding so many bugs while doing it that I spend most of the time fixing them so I can go on.

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
@whirm @devos50 Smooth sailing now! Ready to merge?

**Comment by devos50** [comment]
@brussee I understand you need this endpoint but I'm afraid that if we merge this endpoint, the tests will become too unstable. Also, I first want a review from @whirm ;)

**Comment by devos50** [comment]
@whirm can you review?

File: Tribler/Core/Modules/restapi/downloads_endpoint.py [hunk 3]
```diff
@@ -169,6 +208,53 @@ def render_DELETE(self, request):
...
+        - anon_hops: the number of hops for the anonymous download. 0 hops is equivalent to a plain download
+        - safe_seeding: whether the seeding of the download should be anonymous or not (0 = off, 1 = on)
+        - destination: the download destination path of the torrent
+
+            **Example request**:
+
+                .. sourcecode:: none
+
+                    curl -X PUT http://localhost:8085/download/4344503b7e797ebf31582327a5baae35b11bda01
+                    --data "anon_download=true&safe_seeding=true&destination=/my/dest/on/disk/"
```
Comment (whirm) [inline Tribler/Core/Modules/restapi/downloads_endpoint.py]: We talked about using the hop number instead (0 hops being plain download)

Comment (devos50) [inline Tribler/Core/Modules/restapi/downloads_endpoint.py]: @whirm ah I forgot to update the doc string here. Is that your only comment regarding this PR? :)

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
success :)

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
@whirm can you review this already? (see my last two commits)

**Comment by whirm** [comment]
Looks fine, let's see if the tests pass

**Comment by whirm** [comment]
Also, please send the fixes to next

**Comment by devos50** [comment]
Still the same error on win64...

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
This PR is superseded by #2464

**Comment by whirm** [comment]
Shouldn't it be closed then?

**Comment by brussee** [comment]
@devos50 You can close this one since your awesome commits are merged with #2464. Nice teamwork!

**Pull request closed without merging**